packages/firestore/src/core/query.ts

```diff
@@ -762,6 +799,27 @@ export class KeyFieldInFilter extends FieldFilter {
   }
 }
 
+/** Filter that matches on key fields not present within an array. */
+export class KeyFieldNotInFilter extends FieldFilter {
+  private readonly keys: DocumentKey[];
+
+  constructor(field: FieldPath, value: api.Value) {
+    super(field, Operator.NOT_IN, value);
+    debugAssert(isArray(value), 'KeyFieldNotInFilter expects an ArrayValue');
+    this.keys = (value.arrayValue.values || []).map(v => {
+      debugAssert(
+        isReferenceValue(v),
+        'Comparing on key with NOT_IN, but an array value was not a ReferenceValue'
+      );
+      return DocumentKey.fromName(v.referenceValue);
+    });
+  }
+
+  matches(doc: Document): boolean {
+    return !this.keys.some(key => key.isEqual(doc.key));
```

Comment:
The difference between these types is the operator name and this exclamation point. It seems like you could either:

  * extract a common class for this that takes the operator as a parameter;
  * extract a helper function for creating the keys array;
  * make `KeyFiieldInFilter` the base class and have it take the operator as a parameter;
  * move the construction of the keys array into the caller (passing it as an extra argument) since you now have a dedicated `createKeyFieldInFilter`.

(listed in order of my decreasing preference, but my preference isn't strong.)

Just duplicating the code seems size unfriendly.

Reply:
Extracted into helper.